include/mumble/Base64.hpp

```diff
@@ -3,31 +3,175 @@
+	}
+
+	static inline bool isValid(const char ch) {
+		try {
+			charToByte(ch);
+		} catch (const std::range_error &) {
+			return false;
+		}
+
+		return true;
+	}
+
+	static inline bool isValid(const StrViewConst in) {
+		for (const auto ch : in) {
+			if (!isValid(ch)) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	static constexpr size_t decodedSize(const StrViewConst in) {
+		size_t size = (in.size() / 4) * 3;
+		// Base64 can have up to 2 pad characters, in order to make the string a multiple of 4.
+		// In order to calculate the exact output size we have to ignore padding.
+		if (auto iter = in.rbegin(); *iter == PAD_CHAR) {
+			size -= *++iter == PAD_CHAR ? 2U : 1U;
+		}
+
+		return size;
+	}
+
+	static constexpr size_t encodedSize(const BufViewConst in) {
+		// ceil((float) a / 3) == (a + 2) / 3, assuming a is an integer.
+		return ((in.size() + 2) / 3) * 4;
+	}
+
+	static constexpr size_t decode(BufView out, StrViewConst in) {
+		if (in.empty()) {
+			return 0;
+		}
+
+		if (in.size() % 4) {
+			throw std::invalid_argument("Input is not a multiple of 4!");
+		}
+
+		const size_t size = decodedSize(in);
+		if (out.size() < size) {
+			throw std::invalid_argument("Insufficient output buffer size!");
+		}
+
+		for (; !in.empty(); out = out.subspan(3), in = in.subspan(4)) {
+			out[0] = (charToByte(in[0]) << std::byte(2)) + ((charToByte(in[1]) & std::byte(0x30)) >> 4);
+
+			if (in[2] == '=') {
+				break;
+			}
+
+			out[1] =
+				((charToByte(in[1]) & std::byte(0x0f)) << std::byte(4)) + ((charToByte(in[2]) & std::byte(0x3c)) >> 2);
+
+			if (in[3] == '=') {
+				break;
+			}
+
+			out[2] = ((charToByte(in[2]) & std::byte(0x03)) << std::byte(6)) + charToByte(in[3]);
+		}
+
+		return size;
+	}
+
+	static constexpr size_t encode(StrView out, BufViewConst in) {
+		if (in.empty()) {
+			return 0;
+		}
+
+		const size_t size = encodedSize(in);
+		if (out.size() < size) {
+			throw std::invalid_argument("Insufficient output buffer size!");
+		}
+
+		for (; !in.empty(); out = out.subspan(4)) {
+			out[0] = byteToChar((in[0] & std::byte(0xfc)) >> 2);
 
-	Base64();
-	virtual ~Base64();
+			if (in.size() >= 2) {
+				out[1] = byteToChar(((in[0] & std::byte(0x03)) << 4) + ((in[1] & std::byte(0xf0)) >> 4));
 
-	virtual explicit operator bool();
+				if (in.size() >= 3) {
+					out[2] = byteToChar(((in[1] & std::byte(0x0f)) << 2) + ((in[2] & std::byte(0xc0)) >> 6));
+					out[3] = byteToChar(in[2] & std::byte(0x3f));
+				} else {
+					out[2] = byteToChar((in[1] & std::byte(0x0f)) << 2);
+					out[3] = PAD_CHAR;
+				}
+			} else {
+				out[1] = byteToChar((in[0] & std::byte(0x03)) << 4);
+				out[2] = PAD_CHAR;
+				out[3] = PAD_CHAR;
+			}
 
-	virtual size_t decode(const BufView out, const BufViewConst in);
-	static size_t encode(const BufView out, const BufViewConst in);
+			// All data blocks are guaranteed to be 3 bytes except the last one.
+			in = in.subspan(std::min(size_t{ 3 }, in.size()));
+		}
```

Comment:
Analogous to the `decode` function, a comment textually explaining how this stuff works would be good.